### PR TITLE
Properly fix summarization skipping (introduce summarization queue table) [BA-6186]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: database/migration/src/main/resources/metadata_changesets/summarization_queue_table.xml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            SELECT
+                METADATA_JOURNAL_ID
+            FROM
+                METADATA_ENTRY
+            WHERE
+                METADATA_JOURNAL_ID >= (SELECT SUMMARY_POSITION FROM SUMMARY_STATUS_ENTRY WHERE upper(SUMMARY_NAME) = 'WORKFLOW_METADATA_SUMMARY_ENTRY_INCREASING')
+        </sql>
+    </changeSet>
+
+    <changeSet id="summary_queue_initialization_postgresql" author="gsterin" dbms="postgresql">
```
Comment (mcovarr) [inline database/migration/src/main/resources/metadata_changesets/summarization_queue_table.xml]: ooc is a separate Postgres changeset necessary with `objectQuotingStrategy="QUOTE_ALL_OBJECTS"`?

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 2]
```diff
@@ -178,18 +187,18 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
...
-      previousMetadataEntryIdOption <- getSummaryStatusEntrySummaryPosition(summarizeNameIncreasing)
-      previousMaxMetadataEntryId = previousMetadataEntryIdOption.getOrElse(-1L)
-      nextMaxMetadataEntryId = previousMaxMetadataEntryId + limit
-      updatedSummaryPosition <- summarizeMetadata(
-        minMetadataEntryId = previousMaxMetadataEntryId + 1L,
-        maxMetadataEntryId = nextMaxMetadataEntryId,
+
+      metadataEntryIdsToSummarize <- fetchMetadataJournalIdsFromSummaryQueue(limit.toLong)
+
+      metadataEntriesToSummarize <- dataAccess.metadataEntriesForIds(metadataEntryIdsToSummarize).result
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: TOL feels like these two lines could be done in one shot?

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 3]
```diff
@@ -269,6 +264,40 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
...
+      startMetadataKey, endMetadataKey, nameMetadataKey, statusMetadataKey, submissionMetadataKey, parentWorkflowIdKey, rootWorkflowIdKey)
+    val startsWithMetadataKeys = Set(labelMetadataKey)
+
+    val metadataEntries = rawMetadataEntries filter { entry =>
+      entry.callFullyQualifiedName.isEmpty && entry.jobIndex.isEmpty && entry.jobAttempt.isEmpty &&
+        (exactMatchMetadataKeys.contains(entry.metadataKey) || startsWithMetadataKeys.exists(entry.metadataKey.startsWith))
+    }
+
+    val metadataWithoutLabels = metadataEntries
+      .filterNot(_.metadataKey.contains(labelMetadataKey)) // Why are these "contains" while the filtering is "starts with"?
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: 🙈

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala [hunk 4]
```diff
@@ -137,6 +137,13 @@ trait MetadataEntryComponent {
     }
   )
 
+  def metadataEntriesForIds(ids: Seq[Long]) = {
+    for {
+      entry <- metadataEntries
+      if entry.metadataEntryId.inSetBind(ids)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/MetadataEntryComponent.scala]: TIL

File: database/migration/src/main/resources/metadata_changesets/summarization_queue_table.xml (hunk 1 above)
Comment (grsterin) [reply database/migration/src/main/resources/metadata_changesets/summarization_queue_table.xml]: Yes.  I think that `quotingStrategy` thing doesn't affect raw sql code within the <sql> tag.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 5]
```diff
@@ -269,6 +262,40 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
     runTransaction(action)
   }
 
+  private def summarizeRawMetadata(rawMetadataEntries: Seq[MetadataEntry],
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: This looks like a lot of copy/paste from the `summarizeMetadata` function below. 

I _think_ you should be able to update `summarizeMetadata` to:

* Generate raw metadata in the same way as it does now, but then:
* Pass that set of raw metadata into this function to actually process it

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala [hunk 6]
```diff
@@ -14,5 +15,17 @@ class MetadataDataAccessComponent(val driver: JdbcProfile) extends DataAccessCom
       customLabelEntries.schema ++
       metadataEntries.schema ++
       summaryStatusEntries.schema ++
-      workflowMetadataSummaryEntries.schema
+      workflowMetadataSummaryEntries.schema ++
+      summaryQueueEntries.schema
+
+  // Looks like here is the most appropriate place for this val since it doesn't fit neither in
+  // SummaryQueueEntryComponent nor in MetadataEntryComponent
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala]: This looks like the functions in `SummaryQueueSlickDatabase` to me

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala [hunk 7]
```diff
@@ -14,5 +15,17 @@ class MetadataDataAccessComponent(val driver: JdbcProfile) extends DataAccessCom
...
+  val metadataEntriesToSummarizeQuery = {
+    Compiled(
+      (limit: ConstColumn[Long]) => (for {
+        (_, metadataEntry) <-
+          summaryQueueEntries join metadataEntries on (_.metadataJournalId === _.metadataEntryId)
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala]: Is this join safe in edge cases like "a row exists in the summary queue but not in the metadata entry table"?

**Comment by grsterin** [comment]
> 1. It looks like the perf tests were run on version `"cromwellVersion": "48-e0cee74-SNAP",`, but I don't see that hash in the commit history here.
> 
> I just want to check that was the version you were expecting them to run against, since I would expect it to be a `49-...` hash (you presumably had to rebase onto develop to undo all of the not-quite-summarizer-fix changes)?

@cjllanwarne this is the proper version. I actually took your initial `cjl_summarization_queue` branch and made updates in it. Then I built it locally and pushed to my personal Dockerhub.
I only merged develop branch into this one before creating the PR.

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala (hunk 6 above)
Comment (grsterin) [reply database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala]: That was my initial thought, but i wasn't able to find any single example where we currently mix two different tables in one of those `<TableName>SlickDatabase` classes. Also having it as a function, will cost us benefits of having it as a `Compiled` statement in a val.

**Comment by cjllanwarne** [comment]
> I'm not sure if this would give us some additional performance boost, since we'll need to check each record for matching our criteria.

But sooner or later we have to do that anyway on the summarization side (and probably on the order of 99% of metadata doesn't need summarizing). By doing it before we write to the DB, I think we could avoid:

1. Writing the entries we know don't need to be summarized to the summarization queue.
2. Reading the non-summarizable metadata _value_s from the database into Cromwell (and some of them are pretty big) - only to discard them when we read the key and discover it's not summarizable.

**Comment by grsterin** [comment]
> > I'm not sure if this would give us some additional performance boost, since we'll need to check each record for matching our criteria.
> 
> But sooner or later we have to do that anyway on the summarization side (and probably on the order of 99% of metadata doesn't need summarizing). By doing it before we write to the DB, I think we could avoid:
> 
> 1. Writing the entries we know don't need to be summarized to the summarization queue.
> 2. Reading the non-summarizable metadata _value_s from the database into Cromwell (and some of them are pretty big) - only to discard them when we read the key and discover it's not summarizable.

I'm still reluctant to do that:
1. Yes, but writing a single additional number per entry to a summary table is not nearly a huge overhead, taking into attention that, as you said, in the same transaction we are writing some huge metadata entries into the `metadata_entry` table.
2. Summarizer works asynchronously and though I agree that we should keep it's performance good enough, moving metadata key filtering logic to the "write-metadata-entry" side of things may affect performance of running workflows.
> Reading the non-summarizable metadata _value_s from the database into Cromwell

Also, if my understanding is correct, this is how summarizer works right now.

There are possible things which we may do to optimize summarizer performance, one of which would be to add additional `metadata_key` column to our new queue table, and then allow summarizer to decide if it wants to fetch data from `metadata_entry` table based on that key. But this is food for thought for future optimizations.

File: database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala (hunk 7 above)
Comment (grsterin) [reply database/sql/src/main/scala/cromwell/database/slick/tables/MetadataDataAccessComponent.scala]: I guess it should be fine, since locally I ran summarizer without exceptions against fresh clean `cromwell_test` db.
